src/main/scala/analysis/UtilMethods.scala

```diff
@@ -0,0 +1,52 @@
+package analysis.eval
+import ir.*
+import analysis.solvers.*
+import analysis.Lattice
+import analysis.*
+import util.Logger
+
+/** Evaluate an expression in a hope of finding a global variable.
+  *
+  * @param exp
+  *   : The expression to evaluate (e.g. R1 + 0x1234)
+  * @param n
+  *   : The node where the expression is evaluated (e.g. mem[R1 + 0x1234] <- ...)
+  * @return:
+  *   The evaluated expression (e.g. 0x69632)
+  */
+def evaluateExpression(exp: Expr, n: CfgNode, constantProp: Map[CfgNode, Map[Variable, Any]]): Expr = {
+  Logger.debug(s"evaluateExpression: $exp")
+  exp match {
+    case binOp: BinaryExpr =>
+      val lhs = evaluateExpression(binOp.arg1, n, constantProp)
+      val rhs = evaluateExpression(binOp.arg2, n, constantProp)
+      if (!lhs.isInstanceOf[BitVecLiteral] || !rhs.isInstanceOf[BitVecLiteral]) {
+        return exp
+      }
+      binOp.op match
+        case BVADD  => BitVectorEval.smt_bvadd(lhs.asInstanceOf[BitVecLiteral], rhs.asInstanceOf[BitVecLiteral])
+        case BVSUB  => BitVectorEval.smt_bvsub(lhs.asInstanceOf[BitVecLiteral], rhs.asInstanceOf[BitVecLiteral])
+        case BVASHR => BitVectorEval.smt_bvashr(lhs.asInstanceOf[BitVecLiteral], rhs.asInstanceOf[BitVecLiteral])
+        case BVCOMP => BitVectorEval.smt_bvcomp(lhs.asInstanceOf[BitVecLiteral], rhs.asInstanceOf[BitVecLiteral])
+        case _      => throw new RuntimeException("Binary operation support not implemented: " + binOp.op)
```

Comment:
If casting like this is necessary there should at least be new variables created instead of casting `lhs` and `rhs` on every use.

Reply:
Fixed [Upcoming merge from yousif-memory-region-analysis]